taglib/wavpack/wavpackproperties.cpp

```diff
@@ -212,25 +253,126 @@ void WavPack::Properties::read(File *file, long streamLength)
 
 unsigned int WavPack::Properties::seekFinalIndex(File *file, long streamLength)
 {
-  const long offset = file->rfind("wvpk", streamLength);
-  if(offset == -1)
-    return 0;
+  long offset = streamLength;
+
+  while (offset >= 32) {
+    offset = file->rfind("wvpk", offset - 32);
+
+    if(offset == -1)
+      return 0;
+
+    file->seek(offset);
+    const ByteVector data = file->readBlock(32);
+    if(data.size() < 32)
+      return 0;
+
+    const int version = data.toShort(8, false);
+    if(version < MIN_STREAM_VERS || version > MAX_STREAM_VERS)
+      return 0;
 
-  file->seek(offset);
-  const ByteVector data = file->readBlock(32);
-  if(data.size() < 32)
-    return 0;
+    const unsigned int blockIndex   = data.toUInt(16, false);
+    const unsigned int blockSamples = data.toUInt(20, false);
+    const unsigned int flags        = data.toUInt(24, false);
 
-  const int version = data.toShort(8, false);
-  if(version < MIN_STREAM_VERS || version > MAX_STREAM_VERS)
-    return 0;
+    if (blockSamples && (flags & FINAL_BLOCK))
+      return blockIndex + blockSamples;
+  }
+
+  return 0;
+}
+
+// Given a WavPack block (complete, but not including the 32-byte header), parse the metadata
+// blocks until an ID_SAMPLE_RATE block is found and return the non-standard sample rate
+// contained there, or zero if no such block is found.
+
+int WavPack::Properties::getNonStandardRate(unsigned char const *buffer, int bcount)
+{
+  unsigned char meta_id, c1, c2;
+  int meta_bc;
 
-  const unsigned int flags = data.toUInt(24, false);
-  if(!(flags & FINAL_BLOCK))
-    return 0;
+  while (bcount >= 2) {
+    meta_id = *buffer++;
+    c1 = *buffer++;
 
-  const unsigned int blockIndex   = data.toUInt(16, false);
-  const unsigned int blockSamples = data.toUInt(20, false);
+    meta_bc = c1 << 1;
+    bcount -= 2;
+
+    if (meta_id & ID_LARGE) {
+      if (bcount < 2)
+        return 0;
+
+      c1 = *buffer++;
+      c2 = *buffer++;
+      meta_bc += (static_cast<uint32_t>(c1) << 9) + (static_cast<uint32_t>(c2) << 17);
+      bcount -= 2;
+    }
+
+    if (bcount < meta_bc)
+      return 0;
+
+    // if we got a sample rate, return it
+
+    if ((meta_id & ID_UNIQUE) == ID_SAMPLE_RATE && meta_bc == 4) {
```

Comment:
I do not know much about WavPack, but this seems a bit suspicious. `bcount` is at least `meta_bc`, which is `4`, but `buffer` is incremented here three times, and finally `meta_bc` is added, which gives 7 and this is larger than the guaranteed size of 4. `buffer[3]` would also access beyond the `buffer` if its size is only 4, because at this time, it was already incremented by `3`. Maybe you wanted to access `buffer[0]`, `buffer[1]`, `buffer[2]` instead of the three `*buffer++`?

Reply:
Hi, and thanks for reviewing this! The reason this is okay is that there's an unconditional return at the end of the block. We're guaranteed to have at least 4 bytes and we read the 3 or 4-byte sample rate and return with it because that's all this function does. As you say, the `buffer += meta_bc` would be wrong, but it's never reached. Using `buffer[3]` was also wrong (I copied it from other code without enough context) but fixed it once it failed the test. Of course, that's only for sample rates over 16 MHz, so wouldn't be hit too often!

As for the other fixes, I'll get them in. My C++ is really rusty, but I think I can figure it out (or get someone to help me). Feel free to do it yourself if you like.

Reply:
You are right, there is a `return` statement following the mentioned pointer arithmetic, so it does not matter. My other comments are only minor details, so I think it is OK as it is, and I could do the minor stuff when merging.

Reply:
Okay, sounds good, and thanks again.

It will be great to get this in!